Add 82599 T3 to the list of supported cards

dpino opened the pull request:
The device id for the 82599 T3 card models is 0x151c, which was not added in the list of supported cards. The driver is working fine. We have a machine with two 82599 T3 cards. I could successfully deploy two VMs, launch snabb traffic and ping both machines.

```
link report:
  19 sent on A_NIC.tx -> A_Virtio.rx (loss rate: 0%)
  19 sent on A_Virtio.tx -> A_NIC.rx (loss rate: 0%)
  19 sent on B_NIC.tx -> B_Virtio.rx (loss rate: 0%)
  19 sent on B_Virtio.tx -> B_NIC.rx (loss rate: 0%)
load: time: 1.00s  fps: 3         fpGbps: 0.000 fpb: 0   bpp: 98   sleep: 100 us
load: time: 1.00s  fps: 4         fpGbps: 0.000 fpb: 0   bpp: 98   sleep: 100 us
load: time: 1.00s  fps: 4         fpGbps: 0.000 fpb: 0   bpp: 98   sleep: 100 us
```
Let me know if you want me to run more tests. What else should I try?

dpino commented:
This is the result of the selftest for the driver:

```
sudo SNABB_TEST_INTEL10G_PCIDEVA="0000:04:00.0" \
    SNABB_TEST_INTEL10G_PCIDEVB="0000:04:00.1" \
    $SNABB_SRC/snabb snsh -t apps.intel.intel_app
```

```
-------
Send a bunch of packets from Am0
half of them go to nicAm1 and half go nowhere
link report:
                   0 sent on nicAm0.tx -> sink_ms.in1 (loss rate: 0%)
           3,742,877 sent on nicAm1.tx -> sink_ms.in2 (loss rate: 0%)
           7,486,290 sent on repeater_ms.output -> nicAm0.rx (loss rate: 0%)
                   2 sent on source_ms.out -> repeater_ms.input (loss rate: 0%)
-------
Transmitting bidirectionally between nicA and nicB
link report:
                   0 sent on nicA.tx -> sink.in1 (loss rate: 0%)
                   0 sent on nicB.tx -> sink.in2 (loss rate: 0%)
           3,874,215 sent on source1.out -> nicA.rx (loss rate: 0%)
           3,874,215 sent on source2.out -> nicB.rx (loss rate: 0%)
sq_sq: missing packets
```

lukego commented:
@dpino That looks like a failed test to me. Packets are being sent but not received. Do you have the two ports cabled together? I believe this is expected by the current selftest function.

In your tests with VMs are you sending packets over the wire (physical cable) or are both VMs attached to the same NIC? If the latter then the test might work even if the NIC is not able to transmit/receive on the physical link.

sleinen commented:
You mean 82599 rather than 82559, right?

eugeneia commented:
You should try to run the NFV selftest:

```
cd src; sudo TESTPCI=<pci_address> TELNET_PORT0=5000 TELNET_PORT1=5001 ./program/snabbnfv/selftest.sh`
```

This requires `tmux`, `wget`, `make`, `telnet` and `numactl` in the `PATH`.

dpino commented:
@lukego When I run that test the NICs were cabled together 04:00.0 to 05:00.0. We changed it and cabled the ports of 04:00.0 together, but got the same result while testing the card.

```
Send a bunch of packets from Am0
half of them go to nicAm1 and half go nowhere
link report:
                   0 sent on nicAm0.tx -> sink_ms.in1 (loss rate: 0%)
           4,333,922 sent on nicAm1.tx -> sink_ms.in2 (loss rate: 0%)
           8,668,514 sent on repeater_ms.output -> nicAm0.rx (loss rate: 0%)
                   2 sent on source_ms.out -> repeater_ms.input (loss rate: 0%)
-------
Transmitting bidirectionally between nicA and nicB
link report:
                   0 sent on nicA.tx -> sink.in1 (loss rate: 0%)
                   0 sent on nicB.tx -> sink.in2 (loss rate: 0%)
           4,813,125 sent on source1.out -> nicA.rx (loss rate: 0%)
           4,813,125 sent on source2.out -> nicB.rx (loss rate: 0%)
```

@sleinen Right, fixed :)

@eugenia

Here is the result of running the SnabbNFV selftest.

```
sudo TESTPCI=0000:04:00.0 TELNET_PORT0=6060 TELNET_PORT1=6061 program/snabbnfv/selftest.sh
```

```
Waiting for VM listening on telnet port 6060 to get ready... [OK]
Waiting for VM listening on telnet port 6061 to get ready... [OK]
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/same_vlan.ports
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
[  3]  0.0-10.0 sec  5.16 GBytes  4.43 Gbits/sec
IPERF succeded.
1 packets transmitted, 1 received, 0% packet loss, time 0ms
JUMBOPING succeded.
[  3]  0.0-10.0 sec  12.3 GBytes  10.6 Gbits/sec
IPERF succeded.
tx-checksumming: on
TX-CHECKSUMMING succeded.
tx-checksumming: on
TX-CHECKSUMMING succeded.
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/tx_rate_limit.ports
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
IPERF (RATE_LIMITED) succeded.
IPERF rate is 814 Mbits/sec (900 Mbits/sec allowed).
RATE_LIMITED succeded.
1 packets transmitted, 1 received, 0% packet loss, time 0ms
JUMBOPING succeded.
IPERF (RATE_LIMITED) succeded.
IPERF rate is 814 Mbits/sec (900 Mbits/sec allowed).
RATE_LIMITED succeded.
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/rx_rate_limit.ports
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
IPERF (RATE_LIMITED) succeded.
IPERF rate is 814 Mbits/sec (1200 Mbits/sec allowed).
RATE_LIMITED succeded.
1 packets transmitted, 1 received, 0% packet loss, time 0ms
JUMBOPING succeded.
IPERF (RATE_LIMITED) succeded.
IPERF rate is 814 Mbits/sec (1200 Mbits/sec allowed).
RATE_LIMITED succeded.
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/tunnel.ports
Resolved next-hop fe80::5054:ff:fe00:0 to 52:54:00:00:00:00
Resolved next-hop fe80::5054:ff:fe00:1 to 52:54:00:00:00:01
ND succeded.
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
[  3]  0.0-10.0 sec  10.6 GBytes  9.07 Gbits/sec
IPERF succeded.
1 packets transmitted, 1 received, 0% packet loss, time 0ms
JUMBOPING succeded.
[  3]  0.0-10.0 sec  8.42 GBytes  7.23 Gbits/sec
IPERF succeded.
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/filter.ports
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
Connection to fe80::5054:ff:fe00:0001%eth0 12345 port [tcp/*] succeeded!
PORTPROBE succeded.
Trying ::1...
Connected to localhost.
Escape character is '^]'.
nc -w 1 -q 1 -v  fe80::5054:ff:fe00:0001%eth0 12346
nc: connect to fe80::5054:ff:fe00:0001%eth0 port 12346 (tcp) timed out: Operation now in progress
root@vma:~# 
root@vma:~# Connection closed by foreign host.
FILTER succeded.
USING program/snabbnfv/test_fixtures/nfvconfig/test_functions/stateful-filter.ports
1 packets transmitted, 1 received, 0% packet loss, time 0ms
PING succeded.
Connection to fe80::5054:ff:fe00:0001%eth0 12345 port [tcp/*] succeeded!
PORTPROBE succeded.
Trying ::1...
Connected to localhost.
Escape character is '^]'.
nc -w 1 -q 1 -v  fe80::5054:ff:fe00:0001%eth0 12348
nc: connect to fe80::5054:ff:fe00:0001%eth0 port 12348 (tcp) timed out: Operation now in progress
root@vma:~# 
root@vma:~# Connection closed by foreign host.
FILTER succeded.
Trying ::1...
Connected to localhost.
Escape character is '^]'.
nc -w 1 -q 1 -v  fe80::5054:ff:fe00:0000%eth0 12340
nc: connect to fe80::5054:ff:fe00:0000%eth0 port 12340 (tcp) timed out: Operation now in progress
root@vma:~# 
root@vma:~# Connection closed by foreign host.
FILTER succeded.
```

eugeneia commented:
@dpino Success on the NFV selftest side! I am glad to see its easy to run the NFV selftest now (assuming it just ran out of the box for you).

lukego commented:
So do I interpret correctly that:

1. NFV selftest passes and card is successfully passing traffic between locally attached VMs.
2. NIC selftest is failing and card is not successfuly passing traffic between physical ports.

In this case perhaps the most appropriate thing would be to add support but print a loud warning?

dpino commented:
@lukego That's correct. I run the tests again to confirm it. I updated the PR printing out a warning in case device "0x151c" (82599 T3) is selected.

lukego commented:
I think it would be better to have a more explicit error message saying that the T3 works for loopback traffic but not physical traffic. Then users will know what is expected to work and what is not. Could even print this in the `intel10g` device initialization perhaps?

Also: Please add commits to the PR instead of rebasing. I know, I have asked people to rebase in the past, but I am taking that back because I find that it makes Github based code review really hard and punishes people who merge PRs into their own trees for testing (which is a behavior we want to encourage).  See [github-workflow.md](https://github.com/SnabbCo/snabbswitch/blob/master/src/doc/git-workflow.md) for latest humble requests of contributors :).

We should also fix the driver. @javierguerragiraldez didn't you have this card working at some point? (IIRC it just needed a little time to initialize or something?)

lukego commented:
I am reluctant to merge this. Based on Javier's comments it seems like we could add full support for the card without much more effort than marking it as broken/experimental.

@javierguerragiraldez Could we possibly persuade you to send a new PR that adds support for this card? Based on your [comments](https://groups.google.com/d/msg/snabb-devel/d9sdUAtO8SY/yhdCghzwBgAJ) it sounds to me like the NIC does work but the selftest needs to pause before testing for the link to be usable? However, there are a lot more code changes than this on your 10GbT branch so it would really help if you can strip this down to the minimum needed.

lukego commented:
One idea would be that the selftest could have a "warm up" period. That is, if the link is not actually usable when it reports "link up" then the first step of selftest could be to loop until a packet is actually delivered (or timeout) before proceeding to the real tests. Reasonable?

dpino commented:
@lukego OK, I will give it a try.

eugeneia commented:
> One idea would be that the selftest could have a "warm up" period. That is, if the link is not actually usable when it reports "link up" then the first step of selftest could be to loop until a packet is actually delivered (or timeout) before proceeding to the real tests.

I want to chime in and just suggest we document the eventual "workaround" really well. Like a big flashy comment plus explaining it in the commit message. I remember debugging NIC drivers as being not fun, and think we had issues before when our 10G driver code-rotted slightly. I am very fond of the strict intel selftest we currently have a exactly because of that. Maybe even even do a `ifdef 82599 T3` dependent relaxation?

javierguerragiraldez commented:
I've done some tests on the current `next` branch (which at the moment is the same as the just released 'mango') on Davos `89:00.0`/`.1` card:

- `snabb snsh -t apps.intel.intel_app` passes without modification, but that's because the first tests are loopback, and take some time (100 restarts, 100 reconfigs...)  when it tries to send packets, the card seems to be already warmed up.

- commenting out the three first tests makes the fourth one (`sq_sq()`) fail on 89:00 but it still passes on the SFP+ cards (i.e. 86:00)

- simply adding a 2 sec delay between `engine.configure(c)` and `engine.main({...})` makes it pass again.  1.5 sec seems not to be enough, 2.5 doesn't make it better.

dpino commented:
thanks @javierguerragiraldez I tried the same steps on our machine with the T3 card and I got the following results (tested branch is 'next'):

Current state:

```
Send a bunch of packets from Am0
half of them go to nicAm1 and half go nowhere
link report:
                   0 sent on nicAm0.tx -> sink_ms.in1 (loss rate: 0%)
           4,284,632 sent on nicAm1.tx -> sink_ms.in2 (loss rate: 0%)
           8,569,785 sent on repeater_ms.output -> nicAm0.rx (loss rate: 0%)
                   2 sent on source_ms.out -> repeater_ms.input (loss rate: 0%)
-------
Transmitting bidirectionally between nicA and nicB
link report:
                   0 sent on nicA.tx -> sink.in1 (loss rate: 0%)
                   0 sent on nicB.tx -> sink.in2 (loss rate: 0%)
           4,811,340 sent on source1.out -> nicA.rx (loss rate: 0%)
           4,811,340 sent on source2.out -> nicB.rx (loss rate: 0%)
sq_sq: missing packets
```

`sq_sq` fails and `mq_sq` is not being executed. If I add a 2 second pause between `sq_sq`'s `configure` and `main`:

```
Transmitting bidirectionally between nicA and nicB
link report:
           1,336,576 sent on nicA.tx -> sink.in1 (loss rate: 0%)
           1,336,576 sent on nicB.tx -> sink.in2 (loss rate: 0%)
           2,663,220 sent on source1.out -> nicA.rx (loss rate: 0%)
           2,663,220 sent on source2.out -> nicB.rx (loss rate: 0%)
-------
Send traffic from a nicA (SF) to nicB (two VFs)
The packets should arrive evenly split between the VFs
link report:
                   0 sent on nicAs.tx -> sink_ms.in1 (loss rate: 0%)
                   0 sent on nicBm0.tx -> sink_ms.in2 (loss rate: 0%)
                   0 sent on nicBm1.tx -> sink_ms.in3 (loss rate: 0%)
           8,574,885 sent on repeater_ms.output -> nicAs.rx (loss rate: 0%)
                   2 sent on source_ms.out -> repeater_ms.input (loss rate: 0%)
mq_sq: missing packets
```

`sq_sq` runs successfully but `mq_sq` fails. It seems `mq_sq` needs a 2 second pause too:

```
Send traffic from a nicA (SF) to nicB (two VFs)
The packets should arrive evenly split between the VFs
link report:
                   0 sent on nicAs.tx -> sink_ms.in1 (loss rate: 0%)
           1,731,322 sent on nicBm0.tx -> sink_ms.in2 (loss rate: 0%)
           1,731,364 sent on nicBm1.tx -> sink_ms.in3 (loss rate: 0%)
           3,463,410 sent on repeater_ms.output -> nicAs.rx (loss rate: 0%)
                   2 sent on source_ms.out -> repeater_ms.input (loss rate: 0%)
selftest: ok
```

javierguerragiraldez commented:
> sq_sq runs successfully but mq_sq fails. It seems mq_sq needs a 2 second pause too:

right.  it's because some of these tests not only reconfigure the app network but also reset the NIC.

dpino commented:
I pushed a new commit that adds a time sleep before running tests 'sq_sq' and 'mq_sq'. The 2 second wait only runs if the card model is Intel 82599 T3.

lukego commented:
@dpino I "did a Max" and sent a PR to your PR branch. Could you consider merging those changes please?

Generally this implementation seems good enough but I am still a little unclear. @javierguerragiraldez Is the T3 NIC reporting "link up" before it is able to send traffic? and another 2 seconds are needed? is the SNMP object that reports link status going to have a sane value? or is the issue that we are not trying hard enough to detect when the link is up before running the test?

The 2 second sleep seems fragile and likely to break in the future e.g. when testing other variants of the Intel NIC or with optical transceivers installed in the SFP+ NICs. This seems manageable though: it is only the selftest function that is affected and not the driver itself.

Generally I don't think it is part of the driver contract to be able to pass packets immediately upon startup (and in real life e.g. if we are cabled to a switch running Spanning Tree Protocol then it could take tens of seconds after link-up to actually be able to send traffic). So the selftest requirement of avoiding attempts to send before the link is ready is somewhat unusual.

dpino commented:
@lukego: Sure, merged :)
